Hook up non_differentiability in derivatives.yaml when no autograd function is generated.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19431 Get rid of unnecessary matches_jit_signature: True specifications.
* #19430 Make one_hot non-differentiable.
* #19429 Remove 'BoolTensor', 'IndexTensor' from frontend specifications.
* #19428 Have _embedding_bag_dense_backward match JIT signature.
* #19427 Have embedding_dense_backward match JIT signature.
* **#19426 Hook up non_differentiability in derivatives.yaml when no autograd function is generated.**
* #19425 Move non_differentiable_arg_names from autograd functions to differentiability_info.
* #19424 Stop generating autograd functions for derivatives.yaml entries that only specify output differentiability.
* #19272 Rename 'not_differentiable' to 'non_differentiable'.

This essentially allows us to move a declaration of "BoolTensor/IndexTensor" to derivatives.yaml by specifying that the argument is non-differentiable.

We do this for _cudnn_rnn_backward and there is no codegen change.  This allows us to remove JIT signature mismatch cases.

Differential Revision: [D15003386](https://our.internmc.facebook.com/intern/diff/D15003386)

gchanan committed:
Hook up non_differentiability in derivatives.yaml when no autograd function is generated.

This essentially allows us to move a declaration of "BoolTensor/IndexTensor" to derivatives.yaml by specifying that the argument is non-differentiable.

We do this for _cudnn_rnn_backward and there is no codegen change.  This allows us to remove JIT signature mismatch cases.

cpuhrsch commented on tools/autograd/load_derivatives.py:
```diff
@@ -383,5 +383,5 @@ def find_info(declaration):
         info = find_info(declaration)
         declaration['derivative'] = info['autograd_fn'] if info else None
-        declaration['non_differentiable_arg_names'] = info['non_differentiable_arg_names'] if info else None
+        declaration['non_differentiable_arg_names'] = info['non_differentiable_arg_names'] if info else []
```
Why have 'non_differentiable_arg_names' and not 'differentiable_arg_names' if we're doing all this work already?

gchanan replied:
It's not a bad idea, but would take a lot more code restructuring than is necessary to accomplish schema unification.

For example, there are multiple args_with_derivatives definitions in gen_variable_type that are different.  You'd have to correct this to get something sensible.

cpuhrsch replied:
Ok, maybe we'll do that down the road then.

gchanan committed:
Update on "Hook up non_differentiability in derivatives.yaml when no autograd function is generated."

Hook up non_differentiability in derivatives.yaml when no autograd function is generated.

This essentially allows us to move a declaration of "BoolTensor/IndexTensor" to derivatives.yaml by specifying that the argument is non-differentiable.

We do this for _cudnn_rnn_backward and there is no codegen change.  This allows us to remove JIT signature mismatch cases.

gh-metadata: pytorch pytorch 19426 gh/gchanan/19/head

gchanan committed:
Update on "Hook up non_differentiability in derivatives.yaml when no autograd function is generated."

Hook up non_differentiability in derivatives.yaml when no autograd function is generated.

This essentially allows us to move a declaration of "BoolTensor/IndexTensor" to derivatives.yaml by specifying that the argument is non-differentiable.

We do this for _cudnn_rnn_backward and there is no codegen change.  This allows us to remove JIT signature mismatch cases.

gh-metadata: pytorch pytorch 19426 gh/gchanan/19/head
